consensus/istanbul/proxy/val_enodes_share.go

```diff
@@ -0,0 +1,159 @@
+)
+
+func (pv *proxiedValidatorEngine) generateValEnodesShareMsg(remoteValidators []common.Address) (*istanbul.Message, error) {
+	logger := pv.logger.New("func", "generateValEnodesShareMsg")
+
+	logger.Trace("generateValEnodesShareMsg called", "remoteValidators", common.ConvertToStringSlice(remoteValidators))
+	vetEntries, err := pv.backend.GetValEnodeTableEntries(remoteValidators)
+	logger.Trace("GetValEnodeTableEntries returned", "vetEntries", vetEntries)
+
+	if err != nil {
+		logger.Error("Error in retrieving all the entries from the ValEnodeTable", "err", err)
+		return nil, err
+	}
+
+	sharedValidatorEnodes := make([]sharedValidatorEnode, 0, len(vetEntries))
+	for address, vetEntry := range vetEntries {
+		if vetEntry.GetNode() == nil {
+			continue
+		}
+		sharedValidatorEnodes = append(sharedValidatorEnodes, sharedValidatorEnode{
+			Address:  address,
+			EnodeURL: vetEntry.GetNode().String(),
+			Version:  vetEntry.GetVersion(),
+		})
+	}
+
+	valEnodesShareData := &valEnodesShareData{
+		ValEnodes: sharedValidatorEnodes,
+	}
+
+	valEnodesShareBytes, err := rlp.EncodeToBytes(valEnodesShareData)
+	if err != nil {
+		logger.Error("Error encoding Istanbul Validator Enodes Share message content", "ValEnodesShareData", valEnodesShareData.String(), "err", err)
+		return nil, err
+	}
+
+	msg := &istanbul.Message{
+		Code:      istanbul.ValEnodesShareMsg,
+		Msg:       valEnodesShareBytes,
+		Address:   pv.backend.Address(),
+		Signature: []byte{},
+	}
+
+	logger.Trace("Generated a Istanbul Validator Enodes Share message", "IstanbulMsg", msg.String(), "ValEnodesShareData", valEnodesShareData.String())
+
+	return msg, nil
+}
+
+func (pv *proxiedValidatorEngine) sendValEnodesShareMsg(proxyPeer consensus.Peer, remoteValidators []common.Address) error {
+	logger := pv.logger.New("func", "sendValEnodesShareMsg")
+
+	msg, err := pv.generateValEnodesShareMsg(remoteValidators)
+	if err != nil {
+		logger.Error("Error generating Istanbul ValEnodesShare Message", "err", err)
+		return err
+	}
+
+	// Sign the validator enode share message
+	if err := msg.Sign(pv.backend.Sign); err != nil {
+		logger.Error("Error in signing an Istanbul ValEnodesShare Message", "ValEnodesShareMsg", msg.String(), "err", err)
+		return err
+	}
+
+	// Convert to payload
+	payload, err := msg.Payload()
+	if err != nil {
+		logger.Error("Error in converting Istanbul ValEnodesShare Message to payload", "ValEnodesShareMsg", msg.String(), "err", err)
+		return err
+	}
+
+	logger.Trace("Sending Istanbul Validator Enodes Share payload to proxy peer", "proxyPeer", proxyPeer)
+	if err := proxyPeer.Send(istanbul.ValEnodesShareMsg, payload); err != nil {
+		logger.Error("Error sending Istanbul ValEnodesShare Message to proxy", "err", err)
+		return err
+	}
+
+	return nil
+}
+
+func (pv *proxiedValidatorEngine) SendValEnodesShareMsgToAllProxies() {
+	pv.ph.SendValEnodeShareMsgs()
+}
+
+func (p *proxyEngine) handleValEnodesShareMsg(peer consensus.Peer, payload []byte) (bool, error) {
```

Comment:
i personally prefer to separate code by type vs by trait. Here we are mixing proxyEngine and proxiedValdatorEngine. 
Suggest that we have a file for each type